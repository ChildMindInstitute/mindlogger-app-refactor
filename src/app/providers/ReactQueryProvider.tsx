import { FC, PropsWithChildren } from 'react';

import NetInfo from '@react-native-community/netinfo';
import { createAsyncStoragePersister } from '@tanstack/query-async-storage-persister';
import { QueryClient, onlineManager } from '@tanstack/react-query';
import { PersistQueryClientProvider } from '@tanstack/react-query-persist-client';

import { AsyncStorage, ONE_HOUR } from '@shared/lib';

const queryClient = new QueryClient({
  defaultOptions: {
    mutations: {
      cacheTime: Infinity,
      retry: 0,
    },
    queries: {
      retry: 2,
      cacheTime: ONE_HOUR * 24,
    },
  },
});

const asyncPersist = createAsyncStoragePersister({
  key: 'OFFLINE_CACHE',
  storage: AsyncStorage,
  throttleTime: 1000,
});

onlineManager.setEventListener(setOnline => {
  return NetInfo.addEventListener(state => {
    const status =
      state.isConnected != null &&
      state.isConnected &&
      Boolean(state.isInternetReachable);

    setOnline(status);
  });
});

if (__DEV__) {
  import('react-query-native-devtools').then(({ addPlugin }) => {
    addPlugin({ queryClient });
  });
}

const ReactQueryProvider: FC<PropsWithChildren> = ({ children }) => {
  return (
    <PersistQueryClientProvider
      client={queryClient}
      persistOptions={{
        maxAge: ONE_HOUR * 24,
        persister: asyncPersist,
        buster: 'kill-cache',
      }}
      onSuccess={() =>
        queryClient
          .resumePausedMutations()
          .then(() => queryClient.invalidateQueries())
      }>
      {children}
    </PersistQueryClientProvider>
  );
};

export default ReactQueryProvider;
