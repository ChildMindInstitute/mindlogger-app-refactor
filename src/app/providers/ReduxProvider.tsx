import { FC, PropsWithChildren } from 'react';

import { configureStore } from '@reduxjs/toolkit';
import { Provider } from 'react-redux';

import { IdentityModel } from '@app/entities/identity';

const createDebugger = require('redux-flipper').default;

export const reduxStore = configureStore({
  reducer: { identityReducer: IdentityModel.slices.identityReducer },
  middleware: getDefaultMiddleware => {
    const middlewares = getDefaultMiddleware({ serializableCheck: false });
    middlewares.push(createDebugger());
    return middlewares;
  },
});

const ReduxProvider: FC<PropsWithChildren> = ({ children }) => (
  <Provider store={reduxStore}>{children}</Provider>
);

export default ReduxProvider;
