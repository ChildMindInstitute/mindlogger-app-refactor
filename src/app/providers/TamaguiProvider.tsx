import { FC, PropsWithChildren } from 'react';

import { TamaguiProvider } from '@tamagui/core';

import { appConfig } from '../theme';

const UIProvider: FC<PropsWithChildren> = ({ children }) => (
  <TamaguiProvider config={appConfig}>{children}</TamaguiProvider>
);

export default UIProvider;
