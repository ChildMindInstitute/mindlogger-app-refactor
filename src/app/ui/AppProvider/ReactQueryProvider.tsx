import React, { FC, PropsWithChildren } from 'react';

import NetInfo from '@react-native-community/netinfo';
import { createSyncStoragePersister } from '@tanstack/query-sync-storage-persister';
import { QueryClient, onlineManager } from '@tanstack/react-query';
import { PersistQueryClientProvider } from '@tanstack/react-query-persist-client';

import { createSyncStorage, isAppOnline, useSystemBootUp } from '@shared/lib';

const storage = createSyncStorage('cache-storage');

const queryClient = new QueryClient({
  defaultOptions: {
    mutations: {
      cacheTime: Infinity,
      retry: 0,
    },
    queries: {
      retry: 2,
      cacheTime: Infinity,
      staleTime: Infinity,
    },
  },
});

const syncPersist = createSyncStoragePersister({
  key: 'OFFLINE_CACHE',
  storage,
  throttleTime: 1000,
});

onlineManager.setEventListener(setOnline => {
  return NetInfo.addEventListener(state => {
    const status =
      state.isConnected != null &&
      state.isConnected &&
      Boolean(state.isInternetReachable);

    const mutations = queryClient.getMutationCache().getAll();

    setOnline(status);

    if (mutations.length && status) {
      queryClient.resumePausedMutations();
    }
  });
});

const ReactQueryProvider: FC<PropsWithChildren> = ({ children }) => {
  const { onModuleInitialized } = useSystemBootUp();

  const onCacheRestored = () => {
    isAppOnline().then(isOnline => {
      if (isOnline) {
        onlineManager.setOnline(true);
        queryClient.resumePausedMutations();
      }
    });

    onModuleInitialized('cache');
  };

  return (
    <PersistQueryClientProvider
      client={queryClient}
      persistOptions={{
        maxAge: Infinity,
        persister: syncPersist,
        buster: 'kill-cache',
      }}
      onSuccess={onCacheRestored}
    >
      {children}
    </PersistQueryClientProvider>
  );
};

export default ReactQueryProvider;
