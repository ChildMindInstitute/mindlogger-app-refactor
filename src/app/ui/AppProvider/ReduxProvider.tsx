import { FC, PropsWithChildren } from 'react';

import {
  AnyAction,
  combineReducers,
  configureStore,
  ThunkAction,
} from '@reduxjs/toolkit';
import { Provider } from 'react-redux';
import { persistReducer, persistStore } from 'redux-persist';
import { PersistGate } from 'redux-persist/integration/react';

import { AppletModel } from '@entities/applet';
import { IdentityModel } from '@entities/identity';
import { StreamingModel } from '@entities/streaming';
import { createAsyncStorage, useSystemBootUp } from '@shared/lib';

const storage = createAsyncStorage('redux-storage');

export const persistConfig = {
  key: 'root',
  throttle: 1000,
  storage: storage,
};

const rootReducer = (state: any, action: AnyAction) => {
  if (action.type === 'identity/onLogout') {
    state = undefined;
  }

  const reducer = combineReducers({
    identity: IdentityModel.reducer,
    applets: AppletModel.reducer,
    streaming: StreamingModel.reducer,
  });

  return reducer(state, action);
};

const persistedReducer = persistReducer(
  persistConfig,
  rootReducer,
) as typeof rootReducer;

export const reduxStore = configureStore({
  reducer: persistedReducer,
  middleware: getDefaultMiddleware => {
    const middlewares = getDefaultMiddleware({ serializableCheck: false });

    return middlewares;
  },
});

const persistor = persistStore(reduxStore);

const ReduxProvider: FC<PropsWithChildren> = ({ children }) => {
  const { onModuleInitialized } = useSystemBootUp();

  const onBeforeLift = () => {
    onModuleInitialized('state');
  };

  return (
    <Provider store={reduxStore}>
      <PersistGate persistor={persistor} onBeforeLift={onBeforeLift}>
        {children}
      </PersistGate>
    </Provider>
  );
};

declare global {
  type RootState = ReturnType<typeof reduxStore.getState>;
  type AppDispatch = typeof reduxStore.dispatch;
  type AppThunkAction = ThunkAction<void, RootState, unknown, AnyAction>;
}

export default ReduxProvider;
