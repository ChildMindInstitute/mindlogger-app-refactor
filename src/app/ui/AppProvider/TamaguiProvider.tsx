import { FC, PropsWithChildren } from 'react';

import { TamaguiProvider } from '@tamagui/core';

import { uiConfig } from '@shared/config';

const UIProvider: FC<PropsWithChildren> = ({ children }) => (
  <TamaguiProvider config={uiConfig}>{children}</TamaguiProvider>
);

export default UIProvider;
