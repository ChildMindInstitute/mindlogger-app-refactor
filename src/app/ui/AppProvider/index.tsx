import { FC, Suspense, PropsWithChildren } from 'react';

import { CacheManager } from '@georstat/react-native-image-cache';
import { NavigationContainer } from '@react-navigation/native';
import { Dirs } from 'react-native-file-access';

import CommonEvents from './CommonEvents';
import ReactQueryProvider from './ReactQueryProvider';
import ReduxProvider from './ReduxProvider';
import SplashProvider from './SplashProvider';
import TamaguiProvider from './TamaguiProvider';

CacheManager.config = {
  baseDir: `${Dirs.CacheDir}/images_cache/`,
  blurRadius: 15,
  cacheLimit: 0,
  sourceAnimationDuration: 1000,
  thumbnailAnimationDuration: 1000,
};

const AppProvider: FC<PropsWithChildren> = ({ children }) => {
  return (
    <SplashProvider>
      <ReduxProvider>
        <ReactQueryProvider>
          <TamaguiProvider>
            <NavigationContainer>
              <CommonEvents>
                <Suspense>{children}</Suspense>
              </CommonEvents>
            </NavigationContainer>
          </TamaguiProvider>
        </ReactQueryProvider>
      </ReduxProvider>
    </SplashProvider>
  );
};

export default AppProvider;
