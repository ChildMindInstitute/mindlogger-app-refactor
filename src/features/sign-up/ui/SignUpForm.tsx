/* eslint-disable react-native/no-inline-styles */
import { FC, useMemo, useState } from 'react';
import { TouchableWithoutFeedback } from 'react-native';

import { FormProvider } from 'react-hook-form';
import { useTranslation } from 'react-i18next';
import { isTablet } from 'react-native-device-info';

import {
  executeIfOnline,
  useAppForm,
  useFormChanges,
  colors,
} from '@app/shared/lib';
import { Box, BoxProps, YStack, SubmitButton } from '@shared/ui';
import {
  InputField,
  PasswordRequirements,
  ErrorMessage,
} from '@shared/ui/form';
import { EyeIcon, EyeSlashIcon } from '@shared/ui/icons';

import { SignUpModel } from '../';
import { SignUpFormSchema } from '../validation';

type Props = BoxProps & {
  onLoginSuccess: () => void;
};

const SignUpForm: FC<Props> = props => {
  const { t } = useTranslation();
  const [isPasswordHidden, setPasswordHidden] = useState(true);
  const [displayPasswordRequirements, setDisplayPasswordRequirements] =
    useState(false);

  const {
    isLoading,
    error,
    mutate: signUp,
  } = SignUpModel.useRegistrationMutation(props.onLoginSuccess);

  const { form, submit } = useAppForm(SignUpFormSchema, {
    defaultValues: {
      email: '',
      password: '',
    },
    onSubmitSuccess: data => {
      executeIfOnline(() => signUp(data));
    },
    criteriaMode: 'all',
    shouldUseNativeValidation: false,
  });

  useFormChanges({
    form,
    watchInputs: ['password'],
    onInputChange: function () {
      form.trigger('password');
    },
  });

  const passwordHasLength = !!form.getValues().password.length;
  const passwordRequirements = useMemo(() => {
    const errors = Object.values(
      form.formState.errors?.password?.types || {},
    ).flat();

    return [
      'password_requirements:at_least_characters',
      'password_requirements:no_blank_spaces',
    ].map(key => ({
      label: key,
      isValid: passwordHasLength && !errors.includes(key),
    }));
  }, [passwordHasLength, form.formState.errors?.password?.types]);

  const ShowPasswordIcon = isPasswordHidden ? EyeSlashIcon : EyeIcon;

  return (
    <Box {...props}>
      <FormProvider {...form}>
        <YStack space={isTablet() ? 10 : 22}>
          <InputField
            name="email"
            accessibilityLabel="signup-email-input"
            placeholder={t('auth:email')}
          />

          <InputField
            name="firstName"
            accessibilityLabel="signup-first-name-input"
            placeholder={t('sign_up_form:first_name')}
          />

          <InputField
            name="lastName"
            accessibilityLabel="signup-last-name-input"
            placeholder={t('sign_up_form:last_name')}
          />

          <InputField
            name="password"
            accessibilityLabel="signup-password-input"
            placeholder={t('auth:password')}
            secureTextEntry={isPasswordHidden}
            rightIcon={
              <TouchableWithoutFeedback
                onPress={() => setPasswordHidden(!isPasswordHidden)}
              >
                <ShowPasswordIcon size={24} color={colors.white} />
              </TouchableWithoutFeedback>
            }
            hideError={displayPasswordRequirements || passwordHasLength}
            onFocus={() => setDisplayPasswordRequirements(true)}
            onBlur={() => setDisplayPasswordRequirements(false)}
          />

          {error && (
            <ErrorMessage
              mode="light"
              accessibilityLabel="signup-error-message"
              mt={8}
              error={{ message: error?.evaluatedMessage! }}
            />
          )}

          {(displayPasswordRequirements || passwordHasLength) && (
            <PasswordRequirements requirements={passwordRequirements} />
          )}

          <SubmitButton
            isLoading={isLoading}
            onPress={submit}
            accessibilityLabel="sign_up-button"
            borderRadius={30}
            width="100%"
            bg="$lighterGrey4"
            mt={isTablet() ? 110 : 50}
            textProps={{
              fontSize: 14,
              color: 'black',
            }}
            buttonStyle={{
              alignSelf: 'center',
              paddingVertical: isTablet() ? 13 : 16,
            }}
          >
            {t('sign_up_form:sign_up')}
          </SubmitButton>
        </YStack>
      </FormProvider>
    </Box>
  );
};

export { SignUpForm };
