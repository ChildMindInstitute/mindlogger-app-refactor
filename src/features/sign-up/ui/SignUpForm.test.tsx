import { Provider } from 'react-redux';
import renderer, { act } from 'react-test-renderer';

import ReactQueryProvider from '@app/app/ui/AppProvider/ReactQueryProvider';
import { reduxStore } from '@app/app/ui/AppProvider/ReduxProvider';
import TamaguiProvider from '@app/app/ui/AppProvider/TamaguiProvider';
import { PasswordRequirement } from '@app/shared/ui/form/PasswordRequirements';

import { SignUpForm } from './SignUpForm';

jest.mock('@app/shared/lib/constants', () => ({
  ...jest.requireActual('@app/shared/lib/constants'),
  STORE_ENCRYPTION_KEY: '12345',
}));

jest.mock('@shared/lib', () => ({
  ...jest.requireActual('@shared/lib'),
  useSystemBootUp: jest.fn(() => ({
    onModuleInitialized: jest.fn(),
  })),
}));

jest.mock('@react-native-community/netinfo', () => ({
  addEventListener: jest.fn(),
  fetch: jest.fn(async () => ({
    isConnected: true,
  })),
}));

jest.mock('@app/shared/ui/icons', () => ({
  ...jest.requireActual('@app/shared/ui/icons'),
  FeatherCrossIcon: () => 'x',
  FeatherCheckIcon: () => 'ok',
}));

/* Test helper functions */

const FormTestWrapper = ({ children }: { children: React.ReactNode }) => {
  return (
    <ReactQueryProvider>
      <TamaguiProvider>
        <Provider store={reduxStore}>{children}</Provider>
      </TamaguiProvider>
    </ReactQueryProvider>
  );
};

const createTest = () => {
  const tree = renderer.create(
    <FormTestWrapper>
      <SignUpForm onLoginSuccess={() => {}} />
    </FormTestWrapper>,
  );
  const getRequirements = () => tree.root.findAllByType(PasswordRequirement);
  return { tree, instance: tree.root, getRequirements };
};

describe('SignUp Form', () => {
  it('should render properly', () => {
    const { tree, instance } = createTest();
    act(() => {
      /* display password requirements */
      instance.findByProps({ name: 'password' }).props.onFocus();
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });

  it.todo('should validate password requirements');
});
