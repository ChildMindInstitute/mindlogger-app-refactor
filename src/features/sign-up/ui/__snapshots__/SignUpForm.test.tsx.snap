// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp Form should render properly 1`] = `
<div
  boxSizing="border-box"
  onLoginSuccess={[Function]}
>
  <div
    boxSizing="border-box"
  >
    <div
      boxSizing="border-box"
    >
      <TextInput
        accessibilityLabel="signup-email-input"
        autoCapitalize="none"
        color="rgba(255,255,255,1.00)"
        focusable={true}
        fontSize={20}
        onBlur={[Function]}
        onChangeText={[Function]}
        outlineWidth={0}
        placeholder="auth:email"
        placeholderTextColor="rgba(255,255,255,0.60)"
        selectionColor="black"
        underlineColorAndroid="transparent"
        value=""
      />
    </div>
    <div
      boxSizing="border-box"
      pointerEvents="none"
    />
    <div
      boxSizing="border-box"
    >
      <TextInput
        accessibilityLabel="signup-first-name-input"
        autoCapitalize="none"
        color="rgba(255,255,255,1.00)"
        focusable={true}
        fontSize={20}
        onBlur={[Function]}
        onChangeText={[Function]}
        outlineWidth={0}
        placeholder="sign_up_form:first_name"
        placeholderTextColor="rgba(255,255,255,0.60)"
        selectionColor="black"
        underlineColorAndroid="transparent"
        value=""
      />
    </div>
    <div
      boxSizing="border-box"
      pointerEvents="none"
    />
    <div
      boxSizing="border-box"
    >
      <TextInput
        accessibilityLabel="signup-last-name-input"
        autoCapitalize="none"
        color="rgba(255,255,255,1.00)"
        focusable={true}
        fontSize={20}
        onBlur={[Function]}
        onChangeText={[Function]}
        outlineWidth={0}
        placeholder="sign_up_form:last_name"
        placeholderTextColor="rgba(255,255,255,0.60)"
        selectionColor="black"
        underlineColorAndroid="transparent"
        value=""
      />
    </div>
    <div
      boxSizing="border-box"
      pointerEvents="none"
    />
    <div
      boxSizing="border-box"
    >
      <TextInput
        accessibilityLabel="signup-password-input"
        autoCapitalize="none"
        color="rgba(255,255,255,1.00)"
        focusable={true}
        fontSize={20}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        outlineWidth={0}
        placeholder="auth:password"
        placeholderTextColor="rgba(255,255,255,0.60)"
        secureTextEntry={true}
        selectionColor="black"
        underlineColorAndroid="transparent"
        value=""
      />
      <div
        boxSizing="border-box"
      >
        <Text
          accessible={true}
          allowFontScaling={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          selectable={false}
          style={
            [
              {
                "color": "#FFFFFF",
                "fontSize": 24,
              },
              undefined,
              {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              {},
            ]
          }
        >
          ÔÅ∞
        </Text>
      </div>
    </div>
  </div>
  <span
    fontFamily="System"
    suppressHighlighting={true}
  >
    password_requirements:must_include
  </span>
  <View>
    <View
      testID="password_requirements:at_least_characters"
    >
      <View>
        x
      </View>
      <span
        fontFamily="System"
        suppressHighlighting={true}
      >
        password_requirements:at_least_characters
      </span>
    </View>
    <View
      testID="password_requirements:no_blank_spaces"
    >
      <View>
        x
      </View>
      <span
        fontFamily="System"
        suppressHighlighting={true}
      >
        password_requirements:no_blank_spaces
      </span>
    </View>
  </View>
  <View
    accessibilityLabel="sign_up-button"
    accessibilityState={
      {
        "disabled": false,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <div
      boxSizing="border-box"
    >
      <span
        fontFamily="System"
        suppressHighlighting={true}
      >
        sign_up_form:sign_up
      </span>
    </div>
  </View>
</div>
`;
